src/ewah.zig

```diff
@@ -61,98 +64,208 @@ pub fn ewah(comptime Word: type) type {
+                        source_words[source_index..][0..literal_word_count_chunk],
+                    );
+                    source_index += literal_word_count_chunk;
+                    target_index += literal_word_count_chunk;
+                    decoder_.source_literal_words =
+                        marker.literal_word_count - literal_word_count_chunk;
+                }
+                assert(source_index <= source_words.len);
+                assert(target_index <= target_words.len);
+
+                defer decoder_.target_index = target_index;
+                return target_index - decoder_.target_index;
+            }
+        };
+
+        pub fn decoder(target_words: []Word) Decoder {
+            return .{ .target_words = target_words };
+        }
+
+        // (This is a helper for testing only.)
         /// Decodes the compressed bitset in `source` into `target_words`.
         /// Returns the number of *words* written to `target_words`.
-        // TODO Refactor to return an error when `source` is invalid,
-        // so that we can test invalid encodings.
         pub fn decode(source: []align(@alignOf(Word)) const u8, target_words: []Word) usize {
+            assert(constants.verify);
             assert(source.len % @sizeOf(Word) == 0);
             assert(disjoint_slices(u8, Word, source, target_words));
 
-            const source_words = mem.bytesAsSlice(Word, source);
-            var source_index: usize = 0;
-            var target_index: usize = 0;
-            while (source_index < source_words.len) {
-                const marker: *const Marker = @ptrCast(&source_words[source_index]);
-                source_index += 1;
-                @memset(
-                    target_words[target_index..][0..marker.uniform_word_count],
-                    if (marker.uniform_bit == 1) ~@as(Word, 0) else 0,
-                );
-                target_index += marker.uniform_word_count;
-                stdx.copy_disjoint(
-                    .exact,
-                    Word,
-                    target_words[target_index..][0..marker.literal_word_count],
-                    source_words[source_index..][0..marker.literal_word_count],
-                );
-                source_index += marker.literal_word_count;
-                target_index += marker.literal_word_count;
+            var decoder_ = decoder(target_words);
+            return decoder_.decode(source);
+        }
+
+        pub const Encoder = struct {
+            source_words: []const Word,
+            source_index: usize = 0,
+            /// The number of literals left over from the previous encode() call that still need to
+            /// be copied.
+            literal_word_count: usize = 0,
+
+            /// Returns the number of bytes written to `target_chunk` by this invocation.
+            pub fn encode(encoder_: *Encoder, target_chunk: []align(@alignOf(Word)) u8) usize {
```

Comment:
The reason to have `encode` take a single chunk -- rather than an array of chunks that can be filled in sequence by a single function call -- is that it will let the `CheckpointTrailer` acquire blocks from the grid pool as-needed rather than acquiring the worst-case number. (Leading to fewer evictions from the grid cache.)